keps/30-role-coordination/README.md

```diff
@@ -0,0 +1,270 @@
+
+### Implementation Details
+
+#### API Changes
+
+Add a new `Coordination` field to the RoleBasedGroup spec:
+
+```go
+
+type RoleBasedGroupSpec struct {
+    // Existing fields...
+    
+    // Coordination defines how roles should be coordinated 
+    // +optional
+    Coordination []Coordination `json:"coordination,omitempty"`
+}
+
+type Coordination struct {
+    Strategy []RoleStrategy `json:"strategy,omitempty"`
+}
+
+type RoleStrategy struct {
+    // Role is the name of the role (e.g. "prefill", "decode", "router").
+    Role string `json:"role"`
+    
+    // UpdateStrategy describes how this role should be updated.
+    UpdateStrategy *RoleUpdateStrategy `json:"updateStrategy,omitempty"`
+    
+    // TODO: add more strategy here as needed. e.g. ScalingStrategy
+}
+
+// RoleUpdateStrategy describes how to update a role's workload.
+type RoleUpdateStrategy struct {
+    // Type is the update strategy type (e.g. "Recreate", "InplaceIfPossible").
+    Type string `json:"type,omitempty"`
+    Partition *int32 `json:"partition,omitempty"`
+    MaxUnavailable intstr.IntOrString `json:"maxUnavailable,omitempty"`
+    MaxSurge intstr.IntOrString `json:"maxSurge,omitempty"`
+}
+
+
+```
+
+#### Yaml Example
+
+```yaml
+apiVersion: workloads.x-k8s.io/v1alpha1
+kind: RoleBasedGroup
+metadata:
+  name: role-coordination
+spec:
+  coordination:
+    - strategy:  # strategy 1: reconcile prefill & decode at 2:1 ratio
+        - role: prefill
+          updateStrategy:
+            type: Recreate # [Recreate, InplaceIfPossible]
+            maxSurge: 0
+            partition: 0
+            maxUnavailable: 2
+        - role: decode
+          updateStrategy:
+            type: Recreate # [Recreate, InplaceIfPossible]
+            maxSurge: 0
+            partition: 0
+            maxUnavailable: 1
+    - strategy: # strategy 2: reconcile router & planner at 1:1 ratio
+        - role: router
+          updateStrategy:
+            maxUnavailable: 1
+        - role: planner
+          updateStrategy:
+            maxUnavailable: 1
+```
+
+### Risks and Mitigations
+
+1. **Complexity Risk**: Adding coordination logic increases controller complexity
+    - Mitigation: Implement thorough unit and integration tests
+
+2. **Deadlock Risk**: Poorly configured coordination strategies could cause updates to stall
+    - Mitigation: Add timeouts and clear status reporting
+
+3. **Backward Compatibility**: Existing RoleBasedGroups should continue to work unchanged
+    - Mitigation: Only apply coordination logic when `coordination` is specified
+
+## Design Details
+
+The implementation will modify the main reconciliation loop
+in [RoleBasedGroupReconciler] to check for a coordination strategy. If present, it will execute the coordinated
+update logic; otherwise, it will fall back to the existing independent role update behavior.
+
+Each coordination step will:
+
+1. Apply the specified strategies to the relevant roles
+2. Monitor the status of those roles
```

Comment:
I suggest adding the detail design to instruct the router component to split requests between the old and new Pods in real time during a RoleBasedGroup coordinated upgrade, so that we can work with SGLang Router on rolling update.